src/v/cluster/archival/archival_policy.cc

```diff
@@ -511,6 +512,49 @@ ss::future<candidate_creation_result> archival_policy::get_next_candidate(
     co_return upload;
 }
 
+ss::future<candidate_creation_result> archival_policy::get_next_segment(
+  model::offset begin_inclusive,
+  model::offset end_exclusive,
+  std::optional<model::offset> flush_offset,
+  ss::shared_ptr<storage::log> log,
+  const cloud_storage::partition_manifest& manifest,
+  ss::lowres_clock::duration segment_lock_duration) {
+    std::optional<model::offset> end_inclusive;
+    bool force_upload = upload_deadline_reached();
+    if (flush_offset.has_value()) {
+        end_inclusive = flush_offset.value();
+    }
+    if (force_upload && !end_inclusive.has_value()) {
+        end_inclusive = model::prev_offset(end_exclusive);
+    }
+    vlog(
+      archival_log.debug,
+      "get_next_segment {}, begin_inclusive: {}, end_exclusive: {}, "
+      "end_inclusive: {}, force_upload: {}",
+      _ntp,
+      begin_inclusive,
+      end_exclusive,
+      end_inclusive,
+      force_upload);
+
+    segment_collector segment_collector{
+      begin_inclusive,
+      manifest,
+      *log,
+      config::shard_local_cfg().cloud_storage_segment_size_target().value_or(
+        config::shard_local_cfg().log_segment_size),
+      end_inclusive};
+
+    segment_collector.collect_segments(
+      segment_collector_mode::new_non_compacted);
+    if (!segment_collector.segment_ready_for_upload()) {
+        co_return candidate_creation_error::no_segments_collected;
+    }
+
```

Comment:
I think we need to reset the upload deadline here, right?

Reply:
Right. Good catch!